Cart's Edit modal: supporting product bundles

vmarta opened the pull request:
This PR tweaks the Cart's Edit modal such that it supports the product bundle type. Unlike product set (when added to cart, it shows up as individual items), product bundle shows up as a single-line item. Thus, editing it actually becomes more complex.

The main changes:
- The visual layout of the Edit modal is different enough that it warrants a new component for editing product bundle. But we're still reusing the ProductView component under the hood.
- I took some inspiration from PDP page for some logic like keeping track of the selected variants and validating the orderability of all the child products

Todos:
- [x] change the target branch to `feature/product_bundles`
- [x] delete my console logs
- [x] visual polish
- [x] write tests 
- [ ] manually test with a bundle that contains products without variants

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [x] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- (step1)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

joeluong-sfcc committed:
add initial changes

joeluong-sfcc committed:
implement MVP

joeluong-sfcc committed:
add quantity selection functionality

joeluong-sfcc committed:
refactor and add bundle child product validation

joeluong-sfcc committed:
Refactor code and fix quantity issue

joeluong-sfcc committed:
fix issue where child quantity was not set correctly

joeluong-sfcc committed:
disable add to cart button if child product is unavailable

joeluong-sfcc committed:
refactor PDP

joeluong-sfcc committed:
lint code and go through TODOs

joeluong-sfcc committed:
fix unit tests

joeluong-sfcc committed:
Merge branch 'develop' into feature/bundles_pdp

joeluong-sfcc committed:
add unit tests for pdp and util function

joeluong-sfcc committed:
Merge branch 'feature/bundles_pdp' of github.com:SalesforceCommerceCloud/pwa-kit into feature/bundles_pdp

joeluong-sfcc committed:
update add to cart text for bundles

joeluong-sfcc committed:
add unit tests for product view

joeluong-sfcc committed:
add mock and unit tests for add to cart modal

joeluong-sfcc committed:
address some PR comments

joeluong-sfcc committed:
pass child variant IDs to cart

joeluong-sfcc committed:
Merge branch 'develop' into feature/bundles_pdp

vmarta committed:
Rough attempt at ProductItem supporting bundle and its child variants

joeluong-sfcc committed:
address PR comments

joeluong-sfcc committed:
lint code

joeluong-sfcc committed:
address PR comment

vmarta committed:
Modal's Update button somewhat functional for bundles

vmarta committed:
Hide changes to ProductItem for now

vmarta committed:
Track the selected child products

vmarta committed:
Revert some changes

vmarta committed:
Delete some console logs

vmarta committed:
Allow toggle of ImageGallery

vmarta committed:
Use the correct quantity amount

vmarta committed:
Clarify why hardcoding the quantity

vmarta committed:
Merge branch 'feature/bundles_pdp' into bundles-cart

vmarta committed:
Calculate the correct quantity ourselves

vmarta committed:
Validate orderability before updating the basket item

vmarta committed:
1st attempt at rearranging the components

vmarta committed:
Merge branch 'feature/product_bundles' into bundles-cart

vmarta committed:
Delete some console logs

vmarta committed:
Delete no-longer-necessary code

vmarta committed:
Delete unused references

vmarta committed:
Remove no-longer-needed test

vmarta committed:
Add some comments

vmarta commented on packages/template-retail-react-app/app/components/product-view/index.jsx:
```diff
@@ -347,32 +348,34 @@ const ProductView = forwardRef(
-                            <ImageGallerySkeleton />
-                        )}
-                    </Box>
+                    {showImageGallery && (
```
If `showImageGallery` is true, then this block is rendered. That's all.

vmarta commented on packages/template-retail-react-app/app/hooks/use-product-view-modal.js:
```diff
@@ -73,18 +70,6 @@ export const useProductViewModal = (initialProduct) => {
         }
     }, [])
 
-    useEffect(() => {
-        if (variant) {
-            const {variationValues} = variant
-            // update the url with the new product id and variation values when the variant changes
-            const updatedUrl = rebuildPathWithParams(`${location.pathname}${location.search}`, {
-                ...variationValues,
-                pid: variant.productId
-            })
-            history.replace(updatedUrl)
-        }
-    }, [variant])
```
This is actually no longer necessary, since there's already a similar, newer logic over at ProductView component. When you click on a Swatch, the page url will be updated with the `href` value here: https://github.com/SalesforceCommerceCloud/pwa-kit/blob/22a6aed44f304e2bd048f1e675eb594d2905cc82/packages/template-retail-react-app/app/components/product-view/index.jsx#L447

Furthermore, `pid` url param is not needed on the Cart page anyways.

vmarta commented on packages/template-retail-react-app/app/pages/cart/index.jsx:
```diff
@@ -228,6 +230,44 @@ const Cart = () => {
+                    }
+                })
+            }
+            return await updateItemInBasketMutation.mutateAsync({
```
This API operation results strangely in a 500 server error. It used to be ok.. is there something changed on Joel's ODS instance recently? Maybe this is the result of Jainam in the middle of updating his cartridge?

shethj replied:
Yes. This was me updating the cartridge that was causing 500s

vmarta commented on packages/template-retail-react-app/app/pages/cart/index.jsx:
```diff
@@ -354,7 +394,7 @@ const Cart = () => {
                                         return (
                                             <ProductItem
-                                                key={productItem.productId}
+                                                key={`${productItem.productId}-${idx}`}
```
We can no longer rely on the basket items having distinct product ids. With product bundle, it's now possible to have 2 product bundles appearing as 2 separate line items, while they both have the same product ids.

shethj replied:
We need to use `itemId` or `UUID` instead

joeluong-sfcc replied:
We could also use variant IDs from the children, however, with the way bundles is currently implemented, there would still potential collisions as you can have multiple bundles with the same child variant selections existing as multiple line items

bfeister committed:
fix type coercion error + fix bug in showing `product.price` (aggregate total) where it should be `product.perUnitPrice` (singular total) in edit dialog

bfeister committed:
Fix for price fallback, don't coerce basePrice

joeluong-sfcc commented on packages/template-retail-react-app/app/components/product-view-modal/bundle.jsx:
```diff
@@ -0,0 +1,146 @@
+    onClose: PropTypes.func.isRequired,
+    product: PropTypes.object,
+    isLoading: PropTypes.bool,
+    actionButtons: PropTypes.node,
```
What is `actionButtons` used for? It's defined as a prop but doesn't seem to be used in `product-view-modal/bundle.jsx` or `product-view-modal/index.jsx`

vmarta replied:
Good catch. I think we can safely remove them. I don't see them used anywhere.

joeluong-sfcc commented on packages/template-retail-react-app/app/hooks/use-product-view-modal.test.js:
```diff
@@ -92,30 +92,6 @@ describe('useProductViewModal hook', () => {
         })
     })
 
-    test("update product's related url param when the product content is shown", () => {
```
Would it make sense for this test to move to `app/components/product-view/index.test.js`? I still don't fully understand why the query params get updated on the cart page... Is there a use case or is it just a side effect from how it's handled on the PDP?

bfeister replied:
This is just a side-effect for how state is manage. Even the modal uses query params

vmarta replied:
> Would it make sense for this test to move to app/components/product-view/index.test.js?

I've deleted that test because I've removed the corresponding code here: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1310#discussion_r1241817609

vmarta replied:
> I still don't fully understand why the query params get updated on the cart page

The query params are needed for the swatch selection to fully work. Without the params, users click on the swatches and nothing would happen visually.

Let's contrast it with the PDP. On the PDP, the query params are necessary for these things:
- swatch selection
- AND also for the bookmark-able url (so you can easily share the page that already has its variants chosen)

However, on the Cart page (specifically its Edit modal), the query params are necessary for 1 thing only:
- swatch selection
- (no bookmark-able url for the Cart... that's why the page url would get reset every time you close the modal)

joeluong-sfcc commented on packages/template-retail-react-app/app/components/product-view-modal/bundle.jsx:
```diff
@@ -0,0 +1,146 @@
+                                            showFullLink={false}
+                                            product={product}
+                                            isLoading={isLoading}
+                                            childProductOrderability={childProductOrderability}
```
`childProductOrderability` controls whether or not the `Update` button is disabled or not based on the orderability status of the child products. Since the `Update` button is attached to the parent and not the children, this should get passed as a prop in the parent `ProductView` component

vmarta replied:
I see, thanks! The `childProductOrderability` is for the _parent_ product (specifically to enable/disable its Update button), while the set method for this prop is to be assigned to the _children_ products.

joeluong-sfcc committed:
add unit test

joeluong-sfcc committed:
Merge branch 'feature/product_bundles' into bundles-cart

joeluong-sfcc committed:
add to unit tests

bfeister committed:
lint fix

vmarta commented:
We'll need to address this mobile layout. Maybe in a separate PR?

![Google Chrome 2023-07-03 at 11 44 04](https://github.com/SalesforceCommerceCloud/pwa-kit/assets/847300/dcb8e502-fdde-49f2-8a32-8b9a0cf75f2a)

vmarta committed:
Pass the `childProductOrderability` prop to the correct component

The `childProductOrderability` is for the parent product (specifically to enable/disable its Update button), while the set method for this prop is to be assigned for the children products.

vmarta committed:
Use a better id that's already in the API response

vmarta committed:
Remove no-longer-needed props

vmarta commented on packages/template-retail-react-app/app/components/product-view-modal/bundle.test.js:
```diff
@@ -0,0 +1,98 @@
+    ])
+})
+
+test('renders bundle product view modal', async () => {
```
Thank you for adding these tests!
